### PR TITLE
add annotation tool master branch

### PR DESCRIPTION
**Pull request by StevenPuttemans**
same commit as https://github.com/Itseez/opencv/pull/3653
Video of workings in other PR!

### CODE REVIEW

**Comment by StevenPuttemans** [comment]
Current warning has nothing to do with this code. Lets wait for a rebuild :)

**Comment by mshabunin** [comment]
This tool is not built with OpenCV because there is no corresponding `CMakeLists.txt`.

**Comment by StevenPuttemans** [comment]
@mshabunin I built it myself with OpenCV. Fact is I do not grasp cmake enough to let it build immediatly by the OpenCV installation. If you could guide me through, I could provide a solution for that!

File: apps/annotation/opencv_annotation.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    do
+    {
+        // Keys for processing
+        // You need to select one for confirming a selection and one to continue to the next image
+        // Based on the universal ASCII code of the keystroke: http://www.asciitable.com/
+        //      c = 99		    add rectangle to current image
+        //	    n = 110		    save added rectangles and show next image
+        //	    <ESC> = 27      exit program
+        key_pressed = waitKey(0);
```
Comment (mshabunin) [inline apps/annotation/opencv_annotation.cpp]: Should be `key_pressed = 0xFF & waitKey(0);` to work on Linux.

**Comment by mshabunin** [comment]
I've copied `CMakeLists.txt` from `traincascade` application and modified it accordingly. Also added line to `apps/CMakeLists.txt`.
I guess we need common mechanism for adding apps to OpenCV (similar to `ocv_define_module`) to avoid copy-paste process in future.

File: apps/annotation/opencv_annotation.cpp (hunk 1 above)
Comment (StevenPuttemans) [inline apps/annotation/opencv_annotation.cpp]: Hmm it seems to work on my Ubuntu 14.04. Why should this not work?

**Comment by StevenPuttemans** [comment]
@mshabunin thank you. Can you integrate your additions to my PR? Or will you supply your own PR for those changes?

**Comment by mshabunin** [comment]
I tried it on the 2.4 branch, maybe it can be just copied to master:
https://github.com/mshabunin/opencv/compare/Itseez:2.4...mshabunin:annotate-tool?expand=1
I can make PR myself later, when I will have a little more time.

**Comment by StevenPuttemans** [comment]
I will make the PR, no problem :) Thanks for the info!

**Comment by StevenPuttemans** [comment]
Added it for 2.4 including the fix, will do now for master

https://github.com/Itseez/opencv/pull/3747

**Comment by StevenPuttemans** [comment]
Added it here also ^_^ nice to see this addition getting added, documentation will be for later this week!

**Comment by StevenPuttemans** [comment]
@mshabunin something weird, this branch builds perfectly fine, but the 2.4 one has a bunch of warnings. Will see if I get them fixed there, but no idea why they only happen on 2.4 branch

**Comment by StevenPuttemans** [comment]
@mshabunin fixed everything according to the 2.4 version. This one should merge without problems now :) Had some issues with include orders. Is it normal that you need to include imgcodecs and videoio before imgproc? To me it seems as this shouldn't be allowed to generate problems... however build fails if I switch order.

**Comment by mshabunin** [comment]
@StevenPuttemans, strange, I could not reproduce this problem (Ubuntu 14.04). Can't it be conflict with  library installed in system?

**Comment by StevenPuttemans** [comment]
@mshabunin even weirder ... it doesnt occur anymore here also -_- probably a hickup in one of my installations :D

**Comment by StevenPuttemans** [comment]
Once this gets in I will supply some fixes to the tool on both 2.4 and master. Especially adding more documentation. But for that, I would like this to get merged first.

**Comment by vpisarev** [comment]
:+1:

**Comment by StevenPuttemans** [comment]
Thank you! Updates will come in the next days!

File: apps/annotation/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+project(annotation)
+
+ocv_include_directories("${CMAKE_CURRENT_SOURCE_DIR}" "${OpenCV_SOURCE_DIR}/include/opencv")
+ocv_include_modules(${deps})
+
+set(the_target opencv_annotation)
+
+add_executable(${the_target} opencv_annotation.cpp)
+target_link_libraries(${the_target} ${deps})
```
Comment (alalek) [inline apps/annotation/CMakeLists.txt]: Please use `ocv_add_executable` and `ocv_target_link_libraries` instead (like https://github.com/Itseez/opencv/blob/master/apps/createsamples/CMakeLists.txt).

It is required for OpenCV build with "world" module (it is enabled for distribution package via `-DINSTALL_CREATE_DISTRIB=ON`, see https://github.com/Itseez/opencv/blob/master/CMakeLists.txt#L377).

Comment (StevenPuttemans) [inline apps/annotation/CMakeLists.txt]: Thank you! I will supply a fix in an instant! This will probably be needed for 2.4 branch also?

**Pull request merged**